alacritty/src/input.rs

```diff
@@ -735,6 +745,118 @@ impl<T: EventListener, A: ActionContext<T>> Processor<T, A> {
         }
     }
 
+    /// Handle touch input.
+    pub fn touch(&mut self, touch: TouchEvent) {
+        match touch.phase {
+            TouchPhase::Started => self.on_touch_start(touch),
+            TouchPhase::Moved => self.on_touch_motion(touch),
+            TouchPhase::Ended | TouchPhase::Cancelled => self.on_touch_end(touch),
+        }
+    }
+
+    /// Handle beginning of touch input.
+    pub fn on_touch_start(&mut self, touch: TouchEvent) {
+        let touch_purpose = self.ctx.touch();
+        *touch_purpose = match mem::take(touch_purpose) {
+            Touch::None => Touch::Tap(touch),
+            Touch::Tap(start) => Touch::Zoom(TouchZoom::new((start, touch))),
+            Touch::Zoom(zoom) => Touch::Invalid(zoom.slots()),
+            Touch::Scroll(event) | Touch::Select(event) => {
+                let mut set = HashSet::new();
+                set.insert(event.id);
+                Touch::Invalid(set)
+            },
+            Touch::Invalid(mut slots) => {
+                slots.insert(touch.id);
+                Touch::Invalid(slots)
+            },
+        };
+    }
+
+    /// Handle touch input movement.
+    pub fn on_touch_motion(&mut self, touch: TouchEvent) {
+        let touch_purpose = self.ctx.touch();
+        match touch_purpose {
+            Touch::None => (),
+            // Handle transition from tap to scroll/select.
+            Touch::Tap(start) => {
+                let delta_x = touch.location.x - start.location.x;
+                let delta_y = touch.location.y - start.location.y;
+                if delta_x.abs() > MAX_TAP_DISTANCE {
+                    // Update gesture state.
+                    let start_location = start.location;
+                    *touch_purpose = Touch::Select(*start);
+
+                    // Start simulated mouse input.
+                    self.mouse_moved(start_location);
+                    self.mouse_input(ElementState::Pressed, MouseButton::Left);
+
+                    // Apply motion since touch start.
+                    self.on_touch_motion(touch);
+                } else if delta_y.abs() > MAX_TAP_DISTANCE {
+                    // Update gesture state.
+                    *touch_purpose = Touch::Scroll(*start);
+
+                    // Apply motion since touch start.
+                    self.on_touch_motion(touch);
+                }
```

Comment:
That looks kind of wrong, why you don't compute the actual distance between the points and apply some logic based on that, what's the purpose of separate `x` first and then maybe `y` second handling? Simply to have `Scroll` separate?

Maybe we might use `2` fingers for scrolling, or other terminals scroll with just one finger?

Reply:
Twofingerscrolling on a touchscreen seems pretty unergonomic. I quickly checked some terminals, most don't respond to touch at all, but at least konsole also scrolls with one finger. And of course browsers scroll with one finger.